cpp/include/cuspatial/experimental/detail/derive_trajectories.cuh

```diff
@@ -0,0 +1,156 @@
+                        IdOutputIt ids_out_first,
+                        PointOutputIt points_out_first,
+                        TimestampOutputIt timestamps_out_first,
+                        rmm::cuda_stream_view stream,
+                        rmm::mr::device_memory_resource* mr)
+{
+  using id_type        = detail::iterator_value_type<IdInputIt>;
+  using timestamp_type = detail::iterator_value_type<TimestampInputIt>;
+  using tuple_type     = thrust::tuple<id_type, timestamp_type>;
+
+  auto keys_first     = thrust::make_zip_iterator(ids_first, timestamps_first);
+  auto keys_out_first = thrust::make_zip_iterator(ids_out_first, timestamps_out_first);
+
+  std::size_t temp_storage_bytes = 0;
+  cub::DeviceMergeSort::SortPairsCopy(nullptr,
+                                      temp_storage_bytes,
+                                      keys_first,
+                                      points_first,
+                                      keys_out_first,
+                                      points_out_first,
+                                      std::distance(ids_first, ids_last),
+                                      trajectory_comparator<tuple_type>{},
+                                      stream);
+
+  auto temp_storage = rmm::device_buffer(temp_storage_bytes, stream, mr);
+
+  cub::DeviceMergeSort::SortPairsCopy(temp_storage.data(),
+                                      temp_storage_bytes,
+                                      keys_first,
+                                      points_first,
+                                      keys_out_first,
+                                      points_out_first,
+                                      std::distance(ids_first, ids_last),
+                                      trajectory_comparator<tuple_type>{},
+                                      stream);
+
+  stream.synchronize();
+}
+
+}  // namespace detail
+
+template <typename IdInputIt,
+          typename PointInputIt,
+          typename TimestampInputIt,
+          typename IdOutputIt,
+          typename PointOutputIt,
+          typename TimestampOutputIt,
+          typename OffsetType>
+std::unique_ptr<rmm::device_uvector<OffsetType>> derive_trajectories(
+  IdInputIt ids_first,
+  IdInputIt ids_last,
+  PointInputIt points_first,
+  TimestampInputIt timestamps_first,
+  IdOutputIt ids_out_first,
+  PointOutputIt points_out_first,
+  TimestampOutputIt timestamps_out_first,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  detail::order_trajectories(ids_first,
+                             ids_last,
+                             points_first,
+                             timestamps_first,
+                             ids_out_first,
+                             points_out_first,
+                             timestamps_out_first,
+                             stream,
+                             mr);
+
+  auto const num_points = std::distance(ids_first, ids_last);
```

Comment:
That's not possible -- we only get a `last` iterator for ids.  Assuming (not asserting) the iterator ranges are equal is standard practice in iterator-based APIs like this (and Thrust, and STL).

Reply:
Also, the point of an iterator-based API is flexibility. The caller is free to pass anything that looks like an iterator -- it might be, for example, a counting iterator, or some other dynamic iterator type that generates an open-ended sequence. There is no end. :)